compiler/noirc_frontend/src/elaborator/statements.rs

```diff
@@ -0,0 +1,409 @@
+                    if let Some(definition) = self.interner.try_definition(ident.id) {
+                        mutable = definition.mutable;
+                    }
+
+                    let typ = self.interner.definition_type(ident.id).instantiate(&self.interner).0;
+                    typ.follow_bindings()
+                };
+
+                (HirLValue::Ident(ident.clone(), typ.clone()), typ, mutable)
+            }
+            LValue::MemberAccess { object, field_name, span } => {
+                let (object, lhs_type, mut mutable) = self.elaborate_lvalue(*object, assign_span);
+                let mut object = Box::new(object);
+                let field_name = field_name.clone();
+
+                let object_ref = &mut object;
+                let mutable_ref = &mut mutable;
+                let location = Location::new(span, self.file);
+
+                let dereference_lhs = move |_: &mut Self, _, element_type| {
+                    // We must create a temporary value first to move out of object_ref before
+                    // we eventually reassign to it.
+                    let id = DefinitionId::dummy_id();
+                    let ident = HirIdent::non_trait_method(id, location);
+                    let tmp_value = HirLValue::Ident(ident, Type::Error);
+
+                    let lvalue = std::mem::replace(object_ref, Box::new(tmp_value));
+                    *object_ref =
+                        Box::new(HirLValue::Dereference { lvalue, element_type, location });
+                    *mutable_ref = true;
+                };
+
+                let name = &field_name.0.contents;
+                let (object_type, field_index) = self
+                    .check_field_access(&lhs_type, name, field_name.span(), Some(dereference_lhs))
+                    .unwrap_or((Type::Error, 0));
+
+                let field_index = Some(field_index);
+                let typ = object_type.clone();
+                let lvalue =
+                    HirLValue::MemberAccess { object, field_name, field_index, typ, location };
+                (lvalue, object_type, mutable)
+            }
+            LValue::Index { array, index, span } => {
+                let expr_span = index.span;
+                let (index, index_type) = self.elaborate_expression(index);
+                let location = Location::new(span, self.file);
+
+                let expected = self.polymorphic_integer_or_field();
+                self.unify(&index_type, &expected, || TypeCheckError::TypeMismatch {
+                    expected_typ: "an integer".to_owned(),
+                    expr_typ: index_type.to_string(),
+                    expr_span,
+                });
+
+                let (mut lvalue, mut lvalue_type, mut mutable) =
+                    self.elaborate_lvalue(*array, assign_span);
+
+                // Before we check that the lvalue is an array, try to dereference it as many times
+                // as needed to unwrap any &mut wrappers.
```

Comment:
does this recursive dereferencing overlap with the other one from method calls?

Reply:
They're related in that they're both auto-deref spots Rust has but the one in method calls is much more complex. It includes things we don't have yet such as you're supposed to check if the call is valid at every step before continuing but we just dereference all they way. Then you also often have to add `&mut` implicitly so you dereference down then may have to try building back up.